doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,390 @@
+  * A `DOC ON COLUMN` option naming `v.c1`.
+  * A comment on column `v.c1`.
+  * A `KEY DOC ON TYPE` or `VALUE DOC ON TYPE` option naming `point`.
+  * A `DOC ON TYPE` option naming `point`.
+  * A comment on type `point`.
+
+`c2` is analogous.
+
+For `c1.x`, Materialize will consider the following sources of documentation
+in order:
+
+  * A `KEY DOC ON COLUMN` or `VALUE DOC ON COLUMN` option naming `point.x`.
+  * A `DOC ON COLUMN` option naming `point.x`.
+  * A comment on column `point.x`.
+  * A `KEY DOC ON TYPE` or `VALUE DOC ON TYPE` option naming `integer`.
+  * A `DOC ON TYPE` option naming `integer`.
+  * A comment on type `integer`.
+
+`c1.y` is analogous.
+
+Materialize will also use comments on the materialized view itself as the
+default documentation for the top-level record (named `row` for the key schema
+and `envelope` for the value schema). In the example above, Materialize will
+consider the following sources of documentation in order:
+
+  * A `DOC ON TYPE` comment naming `v`.
+  * A comment on materialized view `v`.
+
+Changes made to comments *after* the the sink is created will not be reflected
+in the Avro schema.
+
+### User-facing documentation
+
+The details in the [Design](#design) section above will need to be worked into
+the Avro sink documentation. In particular, we should be sure to precisely
+explain Avro schema node specifiers and the order in which Materialize searches
+for sources of documentation.
+
+Note that we expect the `COMMENT` interface to be the most common method by
+which users add comments to their sink schemas. The `DOC ON` option is primarily
+useful for the implementation, so that a `CREATE SINK` command can be fully self
+describing. We do expect users to use `DOC ON` in a pinch, e.g. if a single sink
+needs to customize the documentation for a nested type, rather than pulling its
+default description from its comment.
+
+### Implementation
+
+#### Parser
+
+The extensions to the parser are straightforward.
+
+#### Planner
+
+The planner will "freeze" any comments that have been promoted to documentation
+by rewriting the `CREATE SINK` statement to contain those comments as explicit
+`DOC ON` options. This ensures that the Avro schema generated for a sink is
+deterministic—i.e., that future changes to comments on the columns and types
+used by the sink will *not* result in Materialize generating a different schema
+on restart.
+
+#### Catalog
+
+No changes to the catalog are required. The documentation will be automatically
+stored in the catalog via the addition of the `DOC ON` option, as the catalog
+stores the raw DDL statement for each sink.
+
+#### Avro schema generation
+
+The `interchange::json::build_row_schema_json` will need to be adjusted to
+take the requested documentation as a parameter. Loosely, it could grow a
+`docs: BTreeMap<String, String>` parameter, where the key represents the
+field specifier and the value represents the requested documentation for that
+field.
```

Comment:
Good point! @moulimukherjee recommend using named fields for the `Field` variant:

```rust
enum DocTarget {
    Type(GlobalId),
    Field { object_id: GlobalId, column_name: String }
}
```

Also I think it's a `DocTarget` rather than a `DocSource`?

Reply:
sounds good 👍, I will update the doc

Reply:
updated